ClusterProvision

staebler opened the pull request:
* The clusterdeployment controller no longer creates install jobs for clusterdeployments. Instead, the controller creates clusterprovisions. For a given clusterprovision, a single install will be attempted. If the install fails, then the clusterprovision is marked as failed and a new clusterprovision is created.
* The installmanager no longer interacts in any way with clusterdeployments. Instead, the installmanager interacts with clusterprovisions.
* There is no longer any option to specify the hive image to use. Whichever hive image is used by the hive operator will be used by all of the hive players.

dgoodwin commented on pkg/apis/hive/v1alpha1/clusterdeployment_types.go:
```diff
@@ -199,6 +199,10 @@ type ClusterDeploymentStatus struct {
 	// +optional
 	CertificateBundles []CertificateBundleStatus `json:"certificateBundles,omitempty"`
+
+	// Provision is a reference to the last ClusterProvision creating for the deployment
```
"created"

dgoodwin commented on pkg/controller/clusterprovision/clusterprovision_controller.go:
```diff
@@ -0,0 +1,319 @@
+	pLog := log.WithFields(log.Fields{
+		"name":       request.NamespacedName.String(),
+		"controller": controllerName,
+	})
```
Should add reconcile loop metrics similar to the other controllers here. (clusterdeployment is a good example)

dgoodwin commented on pkg/controller/clusterprovision/clusterprovision_controller.go:
```diff
@@ -0,0 +1,319 @@
+		return r.deleteJob(instance, pLog)
+	}
+
+	deployment := &hivev1.ClusterDeployment{}
```
deployment a little unclear here, I'd go with out standard and very concise "cd" var name.

dgoodwin commented on pkg/install/generate.go:
```diff
@@ -197,82 +177,74 @@ func GenerateInstallerJob(
 		Spec: batchv1.JobSpec{
-			Completions:  &completions,
-			BackoffLimit: &backoffLimit,
+			Completions: pointer.Int32Ptr(1),
```
Any benefits to using the Job still or should we switch to run-once Pods?

staebler replied:
The main benefit for using jobs instead of pods is that it is more efficient to watch jobs than to watch all the pods in the cluster as there are far fewer jobs than pods. It is also somewhat easier to ascertain whether a job was successful then whether a pod was successful. I have been torn on whether to scrap jobs in favor of pods, too. Can we put a pin in that and bring it up for discussion and as a candidate for a later PR?

dgoodwin replied:
SGTM

staebler commented:
15897ada...8fc0d126
* Rebased onto master.
* Addressed code review feedback.
* Restored metrics gathering that was removed in the initial work.
* Preserved infra ID to allow future provisions to clean up earlier aborted provisions.

staebler commented:
8fc0d126..e916ae8a
* Use the same hive image for everything. The hive image to use originates with the hive operator.

staebler commented:
```console
$ oc get cd -owide
NAME       CLUSTERNAME   CLUSTERTYPE   BASEDOMAIN                 INSTALLED   INFRAID          AGE
mstaeble   mstaeble                    devcluster.openshift.com   true        mstaeble-dbgnv   33m

$ oc get clusterprovision -owide
NAME                         CLUSTERDEPLOYMENT   STAGE      INFRAID
mstaeble-provision-0-q2vvx   mstaeble            complete   mstaeble-dbgnv
```

staebler commented:
One big thing that I would still like to do is preserve the last failed pod. As it is now, when a provision fails, the job is deleted, along with the pod. This deletes the logs. What I would like to do instead is, when the job fails (as opposed to forcefully deleted by the controller), keep the job from the last provision until the new provision starts. If it is OK, I would like to do that change as a follow-on PR so that I have a chance of landing this PR by the end of the sprint.

staebler commented:
de6ca42e...f5248c32
* Rebased

dgoodwin commented on pkg/controller/clusterprovision/clusterprovision_controller.go:
```diff
@@ -0,0 +1,332 @@
+
+	switch instance.Spec.Stage {
+	case hivev1.ClusterProvisionStageComplete, hivev1.ClusterProvisionStageFailed:
+		return r.deleteJob(instance, pLog)
```
Just saw your comments elsewhere. We can  defer to another PR.

staebler commented:
f5248c32..4b5e18e3
* Ensure that the role for the cluster-installer service account is updated with the new policy rules, even if the role was created by a previous version of hive.

staebler commented on Dockerfile.prebuilt:
```diff
@@ -0,0 +1,13 @@
+COPY --from=builder /go/src/github.com/openshift/hive/bin/hive-operator /opt/services
+
+# TODO: should this be the operator?
+ENTRYPOINT ["/opt/services/manager"]
```
Yeah, these were not even supposed to be in this PR. They were for my testing only. I've been using your `buildah-dev-push` recently anyway. I will remove these artifacts.

staebler commented on Makefile:
```diff
@@ -201,6 +201,16 @@ buildah-dev-push: build
+.PHONY: buildah-push-prebuilt
+buildah-push-prebuilt: buildah-build-prebuilt
+	$(SUDO_CMD) buildah push ${IMG}
+
```
Yes, I will remove this.

staebler commented on hack/e2e-test.sh:
```diff
@@ -100,7 +100,6 @@ go run "${SRC_ROOT}/contrib/cmd/hiveutil/main.go" create-cluster "${CLUSTER_NAME
 	--ssh-public-key-file="${CLOUD_CREDS_DIR}/ssh-publickey" \
 	--pull-secret-file="${CLOUD_CREDS_DIR}/pull-secret" \
 	--base-domain="${BASE_DOMAIN}" \
-	--hive-image="${HIVE_IMAGE}" \
```
I've removed the `--hive-image` parameter completely from the create-cluster command.

staebler commented on pkg/apis/hive/v1alpha1/clusterimageset_types.go:
```diff
@@ -6,12 +6,6 @@ import (
-	// If not present, the default Hive image for the clusterdeployment controller
-	// is used.
-	// +optional
-	HiveImage *string `json:"hiveImage,omitempty"`
```
The APIServer will happily accept it. You can always add whatever extra stuff you want. It will get stripped out by the JSON unmarshal.

dgoodwin commented on pkg/installmanager/installmanager.go:
```diff
@@ -931,7 +876,7 @@ func initSSHAgent(privKeyPath string, m *InstallManager) (func(), error) {
 }
 
-func uploadInstallerLog(cd *hivev1.ClusterDeployment, m *InstallManager, installErr error) error {
+func uploadInstallerLog(provision *hivev1.ClusterProvision, m *InstallManager, installErr error) error {
```
Per earlier comment this is probably the spot we need to add it to the provision and kill the configmap entirely.

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -168,6 +169,14 @@ func AddToManager(mgr manager.Manager, r reconcile.Reconciler) error {
+		OwnerType:    &hivev1.ClusterDeployment{},
+	}); err != nil {
+		return err
+	}
```
Yes, we still need to watch jobs because there are still jobs used for image resolution.

staebler replied:
As for the spec changes kicking off a new provision, we had a conversation about that a few weeks ago on slack. You indicated that it would be acceptable to not stop the current install. Any changes to the spec would be reconciled later on the installed cluster.

See https://coreos.slack.com/archives/CE3ETN3J8/p1562697405437400

dgoodwin commented:
Ok I think that's all I've got. Looks great overall. Still debugging why my cluster's are retrying or even requeuing after a failure.

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -449,156 +485,177 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
-			existingJob = nil
-		} else {
-			cdLog.WithError(err).Error("error looking for install job")
+	provisionName := apihelpers.GetResourceName(cd.Name, fmt.Sprintf("provision-%d-%s", cd.Status.InstallRestarts, utilrand.String(5)))
```
Sounds reasonable.

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -449,156 +485,177 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
 
-	// firstInstalledObserve will be true if this is the first time we've noticed the install job completed.
-	// If true, we know we can report the metrics associated with a completed job.
-	if firstInstalledObserve {
```
That is correct. If we hit this point, that is the first installed observe. If we are setting `cd.Spec.Installed` to true when it was previously false, that is the first installed observe: There is no need for an annotation for that.

staebler commented:
10abaf85..dfb39eff
* Updated e2e tests so that they wait for jobs based upon labels on the jobs rather than expected names for the jobs.

staebler commented:
dfb39eff..a30bb181
* Save installer log and metadata directly in clusterprovision instead of in configmaps.
* Save kubeconfig and admin password at the same time that the metadata is saved. That way we avoid building out the cluster if there is a problem with them.
* Rebased on master.

staebler commented:
1eab7d0f...768d2f47
* Process install logs for failure reason from clusterprovision controller.

staebler commented:
c3d6e063..5b845995
* Do not delete install job when it succeeds or fails. Only delete an install job if the clusterprovision needs to be aborted, since in that case the pod needs to be killed.
* Keep at most 3 clusterprovisions for a given clusterdeployment (plus the initial clusterprovision since that is used to determine when the first installation started).

dgoodwin commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -1564,3 +1587,85 @@ func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd
+func (r *ReconcileClusterDeployment) deleteStaleProvisions(provs []*hivev1.ClusterProvision, cdLog log.FieldLogger) {
+	// Cap the number of existing provisions. Always keep the earliest provision as
+	// it is used to determine the total time that it took to install. Take off
+	// one extra to make room for the new provision being started.
```
Ok this is awesome, I can't remember mentioning it but I thought first + a few most recent would be ideal but too fussy to request.  +100

dgoodwin commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -449,156 +490,185 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
+		nextProvisionTime := calculateNextProvisionTime(provision, cdLog)
+		timeUntilNextProvision := time.Until(nextProvisionTime)
+		if timeUntilNextProvision.Seconds() > 0 {
+			cdLog.WithField("next provision", nextProvisionTime).Info("waiting to start a new provision after failure")
```
nextProvision a little better for the field name, we can't do structured log searching but if we had a decent log stack we should be able to.

dgoodwin replied:
It would be neat if this was stored in a condition on the ClusterDeployment someday just so it's externally viewable to OCM or someone in the event of a cluster failing but not doing anything right now. Not essential for this PR though.

staebler commented on pkg/apis/hive/v1alpha1/clusterprovision_types.go:
```diff
@@ -0,0 +1,133 @@
+	PodSpec corev1.PodSpec `json:"podSpec"`
+
+	// Attempt is which attempt number of the cluster deployment that this ClusterProvision is
+	Attempt int `json:"attempt"`
```
It is attempt 0. All counting starts at 0. I easily can change it to start at 1 instead, if you prefer.

staebler commented on pkg/installmanager/installmanager.go:
```diff
@@ -473,7 +443,7 @@ func (m *InstallManager) generateAssets(cd *hivev1.ClusterDeployment) error {
 	// If the failure test annotation is set, write a bogus resource into the manifests which will
 	// cause a late failure in the install process, enough that we can gather logs from the cluster.
-	if b, err := strconv.ParseBool(cd.Annotations[constants.InstallFailureTestAnnotation]); b && err == nil {
+	if b, err := strconv.ParseBool(provision.Annotations[constants.InstallFailureTestAnnotation]); b && err == nil {
```
Yep. I'll fix that.

staebler replied:
This has been resolved by adding the annotation to the ClusterProvision when it is created.

dgoodwin commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -608,16 +678,45 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
+		cdLog.WithError(err).Error("error looking for install job")
+		return &reconcile.Result{}, err
+	}
+	cdLog.Info("waiting for legacy install job to de removed")
```
to be

dgoodwin commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -449,156 +490,185 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
+		}
+		cdLog.Info("clearing current failed provision to make way for a new provision")
+		return r.clearOutCurrentProvision(cd, cdLog)
+	case hivev1.ClusterProvisionStageComplete:
```
Absolutely not for this PR, but I kinda feel like there's a lot of ClusterProvision processing code in the ClusterDeployment controller. Do you think some of this would be reasonable to refactor out to the ClusterProvision controller in future? Not sure if you feel this is the best place for it, or if it's just how it evolved given the job code it replaced.

staebler replied:
Hmm, I'm not sure that I feel the same way as you. The code around the area of the comment is doing things on the ClusterDeployment, which makes it appropriate for the ClusterDeployment controller. The ClusterProvision controller reconciles ClusterProvisions and should not care about the ClusterDeployment. In fact, you may notice that the ClusterProvision controller does not even get the ClusterDeployment associated with the ClusterProvision.

staebler commented:
d5f17649..d0fc110c
* Create job for provision in one reconcile and adopt the job in the next reconcile.
* Create provision for clusterdeployment in one reconcile and adopt the provision in the next reconcile.
* Add a ProvisionFailed status condition to the clusterdeployment when a provision fails. The message of this condition will include the time when the next provision will be created.
* Delete the outstanding provision when a clusterdeployment is deleted. This will terminate an running provision jobs waiting for them to complete prior to completing the deletion of the clusterdeployment.

staebler commented:
985bae84..6faafc27
* Included permissions for clusterprovisions in the rest of the RBAC roles.
* Added the admission control for validating clusterprovisions to the one started by the hiveadmission.

staebler commented on config/rbac/manager_role.yaml:
```diff
@@ -112,10 +121,9 @@ rules:
-  - dnszones/status
-  - dnszones/finalizers
-  - dnsendpoints
+  - clusterprovisions
```
Thanks. I have added permissions for `clusterprovisions` to all of the roles.

LalatenduMohanty commented:
@staebler This PR contains 40 commits as of now which is a lot IMO. As we have gone through several rounds of review, do you think it is fine now to squash them appropriately to logical ones?

staebler commented:
> @staebler This PR contains 40 commits as of now which is a lot IMO. As we have gone through several rounds of review, do you think it is fine now to squash them appropriately to logical ones?

Sure. I can take some time this morning to squash done to logical commits.

csrwng commented:
/retest

LalatenduMohanty commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -410,22 +435,50 @@ func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hi
+	case err != nil:
+		cdLog.WithError(err).Error("Error updating the merged pull secret")
+		return reconcile.Result{}, err
+	case modifiedCD:
```
This is not related to cluster provision change, so if possible it should be a separate commit

staebler replied:
I will remove this. It is a remnant of when I changed things to create the pull secret anew with every ClusterProvision, not realizing at the time that the pull secret was needed for the image resolution job as well.

staebler commented:
6faafc27...f212ef54
* Collapsed into 10 commits

staebler commented on contrib/cmd/waitforjob/main.go:
```diff
@@ -68,70 +79,83 @@ func (w *waitForJobOpts) Run() {
+		case err == nil:
+			return
+		case err != wait.ErrWaitTimeout:
+			log.WithError(err).Fatal("job existence failed")
```
No. It is a fatal which will terminate the program.

csrwng replied:
Sorry, thx

staebler commented on contrib/cmd/waitforjob/main.go:
```diff
@@ -68,70 +79,83 @@ func (w *waitForJobOpts) Run() {
 
-	if err := w.waitForJobExistence(client, namespace); err != nil {
-		log.WithError(err).Fatal("job existence failed")
+	if w.jobType == uninstallJobType {
```
My understanding is that, in the uninstall case, the wait-for may come too late and miss the uninstall job. That is why it passes when the uninstall job does not exist. The test should never miss the install job as that takes a lot longer to run. Perhaps passing on a missing uninstall job is not necessary any more as the uninstall job should take a significant chunk of time, too.

staebler replied:
So, if anything, neither the install case or the uninstall case should pass when the job does not exist, in my opinion.

csrwng replied:
The reason we wait for the job to exist is that it won’t get created as soon as you perform the action like creating or deleting the cluster deployment. There’s the really small time between the action and the controller syncing the cluster deployment and creating the job. For create, the e2e has a 2 min sleep which seems wrong.

staebler replied:
Oh, I get your point now. The wait-for does not need to wait for the job to exist. The job can be created after the wait-for starts, and it will still be picked up.

staebler commented on pkg/apis/hive/v1alpha1/validating-webhooks/clusterprovision_validating_admission_hook.go:
```diff
@@ -0,0 +1,291 @@
+}
+
+// ValidatingResource is called by generic-admission-server on startup to register the returned REST resource through which the
+//                    webhook is accessed by the kube apiserver.
```
I can fix it here, but that is the whitespace pattern used in all of the validating admission controllers. See https://github.com/openshift/hive/blob/10915d60aad9e26128b942f8bcae9bca06450795/pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go#L68-L71, for example.

csrwng replied:
Then let’s keep it that way if you like. I now realized that we indent to the second word in the first line. It just looks random.

staebler commented on pkg/apis/hive/v1alpha1/validating-webhooks/clusterprovision_validating_admission_hook.go:
```diff
@@ -0,0 +1,291 @@
+		return resp
+	}
+
+	logger = logger.WithField("object.Name", newObject.Name)
```
Yep. That makes sense. I will change that. Note, however, that this too is something that is done this way in all of the validating admission controllers.

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -233,9 +243,10 @@ type ReconcileClusterDeployment struct {
 // +kubebuilder:rbac:groups=hive.openshift.io,resources=clusterimagesets,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=hive.openshift.io,resources=clusterimagesets/status,verbs=get;update;patch
-func (r *ReconcileClusterDeployment) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+func (r *ReconcileClusterDeployment) Reconcile(request reconcile.Request) (result reconcile.Result, returnErr error) {
```
Yes. I am including `result` in the "reconcile complete" info log so that we can see whether the requeueAfter field is what we expect.

https://github.com/openshift/hive/blob/f212ef54ec1a17c06fba131ee74d283edf6302d2/pkg/controller/clusterdeployment/clusterdeployment_controller.go#L259-L263

csrwng replied:
Ack

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -269,10 +280,17 @@ func (r *ReconcileClusterDeployment) Reconcile(request reconcile.Request) (recon
 }
 
-func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (reconcile.Result, error) {
+func (r *ReconcileClusterDeployment) reconcile(request reconcile.Request, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (result reconcile.Result, returnErr error) {
```
I am using `result` in a defer to determine whether the requeueAfter time needs to be adjusted for the delete-after annotation.

https://github.com/openshift/hive/blob/f212ef54ec1a17c06fba131ee74d283edf6302d2/pkg/controller/clusterdeployment/clusterdeployment_controller.go#L391-L402

staebler commented on pkg/controller/clusterdeployment/clusterdeployment_controller.go:
```diff
@@ -1030,31 +1109,26 @@ func (r *ReconcileClusterDeployment) syncDeletedClusterDeployment(cd *hivev1.Clu
+			return reconcile.Result{}, nil
+		default:
+			cdLog.Debug("still waiting for outstanding provision to be removed")
+			return reconcile.Result{}, nil
```
I don't think so. My understanding is that the clusterdeployment will be requeued when the clusterprovision is removed. It seems to be working that way, but I can dig into it deeper to verify that that is correct.

csrwng replied:
My understanding is that you get an event when the delete action happens, but not when the resource actually goes away.

staebler replied:
OK. I will verify the behavior and add a requeue if needed.

staebler replied:
You are correct about the behavior. The ClusterDeployment is not requeued when the ClusterProvision is finally removed. It was working in my tests because the ClusterDeployment was being requeued as a result of other changes. I will add a RequeueAfter here.

staebler commented on pkg/controller/clusterprovision/clusterprovision_controller.go:
```diff
@@ -0,0 +1,330 @@
+const (
+	controllerName = "clusterProvision"
+
+	adminSSHKeySecretKey = "ssh-publickey"
```
You are correct. I will remove it.

staebler commented on pkg/controller/clusterprovision/clusterprovision_controller_test.go:
```diff
@@ -0,0 +1,321 @@
+		{
+			name: "running job",
+			existing: []runtime.Object{
+				testProvision(withJob()),
```
@twiest has been adding some to `pkg/controller/test`. We had a discussion about this pattern while you were on sabbatical with the installer team. The idea is to gradually use these more and more throughout the tests.

staebler commented:
f212ef54...c803b0a8
* Addressed a few code review nits.

LalatenduMohanty commented on pkg/controller/utils/image.go:
```diff
@@ -1,23 +0,0 @@
-		return *imageSet.Spec.HiveImage
-	}
-	return images.GetHiveImage(cdLog)
-}
```
This is getting added in https://github.com/openshift/hive/pull/482/commits/9279579d1387e1da2a45b511090e8c7c2407f519. Not sure why it is getting removed from here.

staebler replied:
The idea is that each commit is discrete and leaves the product in a buildable and usable state. In https://github.com/openshift/hive/commit/9279579d1387e1da2a45b511090e8c7c2407f519, the function was supposed to have been moved from ReconcileClusterDeployment to controllerutils. I see that in my restructuring of the commits, I did not remove it from ReconcileClusterDeployment. I will fix that commit to remove it.

LalatenduMohanty commented:
I was trying to run Hive with this PR on minishift but there seems to be some issue
```
hive-controllers-5b56dcc4b6-9czwd   0/1       Error     0          19s
hive-operator-57dc6446df-skd9d      1/1       Running   0          19s
hiveadmission-5dfff7f575-767hm      1/1       Running   0          19s
hiveadmission-5dfff7f575-fvqrr      1/1       Running   0          19s
```
It might be some issue with my setup as I was using the same setup for testing some other PRs. However wanted to findout if it is ready for testing. If yes then I can do further testing. 

```
$ oc describe pod hive-controllers-5b56dcc4b6-9czwd -n hive

xxxxxxxxxxxxxxxxxxxxxxxxxx

Events:
  Type     Reason     Age               From                Message
  ----     ------     ----              ----                -------
  Normal   Scheduled  1m                default-scheduler   Successfully assigned hive/hive-controllers-5b56dcc4b6-9czwd to localhost
  Normal   Pulling    16s (x4 over 1m)  kubelet, localhost  pulling image "quay.io/lalatendum/hive-devel:latest"
  Normal   Pulled     13s (x4 over 1m)  kubelet, localhost  Successfully pulled image "quay.io/lalatendum/hive-devel:latest"
  Normal   Created    13s (x4 over 1m)  kubelet, localhost  Created container
  Normal   Started    13s (x4 over 1m)  kubelet, localhost  Started container
  Warning  BackOff    11s (x6 over 1m)  kubelet, localhost  Back-off restarting failed container
```

staebler commented:
> I was trying to run Hive with this PR on minishift but there seems to be some issue
> 
> ```
> hive-controllers-5b56dcc4b6-9czwd   0/1       Error     0          19s
> hive-operator-57dc6446df-skd9d      1/1       Running   0          19s
> hiveadmission-5dfff7f575-767hm      1/1       Running   0          19s
> hiveadmission-5dfff7f575-fvqrr      1/1       Running   0          19s
> ```
> 
> It might be some issue with my setup as I was using the same setup for testing some other PRs. However wanted to findout if it is ready for testing. If yes then I can do further testing.
> 
> ```
> $ oc describe pod hive-controllers-5b56dcc4b6-9czwd -n hive
> 
> xxxxxxxxxxxxxxxxxxxxxxxxxx
> 
> Events:
>   Type     Reason     Age               From                Message
>   ----     ------     ----              ----                -------
>   Normal   Scheduled  1m                default-scheduler   Successfully assigned hive/hive-controllers-5b56dcc4b6-9czwd to localhost
>   Normal   Pulling    16s (x4 over 1m)  kubelet, localhost  pulling image "quay.io/lalatendum/hive-devel:latest"
>   Normal   Pulled     13s (x4 over 1m)  kubelet, localhost  Successfully pulled image "quay.io/lalatendum/hive-devel:latest"
>   Normal   Created    13s (x4 over 1m)  kubelet, localhost  Created container
>   Normal   Started    13s (x4 over 1m)  kubelet, localhost  Started container
>   Warning  BackOff    11s (x6 over 1m)  kubelet, localhost  Back-off restarting failed container
> ```

@LalatenduMohanty Would you mind looking in the logs for that pod to see if it gives any indication why it is crashing?

LalatenduMohanty commented:
> @LalatenduMohanty Would you mind looking in the logs for that pod to see if it gives any indication why it is crashing?

I have deleted the minishift instance and will try it on a new instance and keep you updated.

LalatenduMohanty commented:
@staebler Things are working fine for me on a fresh setup. 
```
$ oc get pods
NAME                                        READY     STATUS      RESTARTS   AGE
lalatendum-test-2-0-hxxg9-provision-v5xp8   1/3       Running     0          7m
lalatendum-test-2-imageset-42ntn            0/2       Completed   0          7m
```
```
$ oc get pods -n hive
NAME                                READY     STATUS    RESTARTS   AGE
hive-controllers-5b56dcc4b6-882l5   1/1       Running   0          10m
hive-operator-57dc6446df-hqmdz      1/1       Running   0          10m
hiveadmission-5dfff7f575-8szzr      1/1       Running   0          10m
hiveadmission-5dfff7f575-v7x2c      1/1       Running   0          10m
```

LalatenduMohanty commented:
/retest 
As the golangci lint runs by default now.

LalatenduMohanty commented:
/retest
